packages/js/product-editor/src/hooks/use-validation.ts

```diff
@@ -0,0 +1,65 @@
+/**
+ * External dependencies
+ */
+import { useDispatch } from '@wordpress/data';
+import { useEffect, useState } from '@wordpress/element';
+
+/**
+ * Signals that product saving is locked.
+ *
+ * ```typescript
+ * // Sync validation
+ * const validateTitle = useCallback((): boolean => {
+ *  if ( product.title.length < 2 ) {
+ *    return false;
+ *  }
+ *  return true;
+ * }, [product]);
+ * ...
+ * const isTitleValid = useValidation( 'product/title', validateTitle );
+ *
+ * // Async validation
+ * const validateSlug = useCallback(async (): Promise<boolean> => {
+ *  return fetch(`.../${product.id}/validate-slug`)
+ *    .then(response => response.json())
+ *    .then(({ isValid }) => isValid)
+ *    .catch(() => false);
+ * }, [product])
+ * ...
+ * const isSlugValid = useValidation( 'product/slug', validateSlug );
+ * ```
+ *
+ * @param lockName The namespace used to lock the product saving if validation fails.
+ * @param validate The validator function.
+ */
+export function useValidation(
+	lockName: string,
+	validate: () => boolean | Promise< boolean >
+): boolean | undefined {
```

Comment:
Is it possible to limit this to purely `boolean` or does `undefined` add value here?

Reply:
I did that intentionally because we don't want to validate a field if it's not dirty or touched, so I left the `undefined` there to cover those scenarios. If we remove it then we will need other custom hooks to capture dirty/touched states.